src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.DecodeRecipients.cs

```diff
@@ -0,0 +1,306 @@
+            //      recipientEncryptedKeys RecipientEncryptedKeys }
+
+            DerSequenceReader encodedKeyAgreement = recipientSet.ReadSequence();
+            int version = encodedKeyAgreement.ReadInteger();
+
+            // OriginatorIdentifierOrKey::= CHOICE {
+            //      issuerAndSerialNumber IssuerAndSerialNumber,
+            //      subjectKeyIdentifier[0] SubjectKeyIdentifier,
+            //      originatorKey[1] OriginatorPublicKey }
+
+            const byte constructedSequence = (byte)DerSequenceReader.DerTag.Sequence | DerSequenceReader.ConstructedFlag;
+            const byte contextSpecific1 = DerSequenceReader.ConstructedFlag | DerSequenceReader.ContextSpecificTagFlag | 0x01;
+
+            DerSequenceReader originator = encodedKeyAgreement.ReadSequence();
+            byte tag = originator.PeekTag();
+
+            SubjectIdentifierOrKey originatorIdentifierOrKey;
+
+            switch (tag)
+            {
+                case constructedSequence:
+                    X509IssuerSerial issuerSerial = ReadIssuerSerial(originator);
+
+                    originatorIdentifierOrKey =
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.IssuerAndSerialNumber, issuerSerial);
+
+                    break;
+
+                case DerSequenceReader.ContextSpecificTagFlag:
+                    string skid = ReadSubjectKeyId(originator);
+
+                    originatorIdentifierOrKey =
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.SubjectKeyIdentifier, skid);
+
+                    break;
+
+                case contextSpecific1:
+                    PublicKeyInfo publicKeyInfo = ReadOriginatorPublicKey(originator);
+
+                    originatorIdentifierOrKey =
+                        new SubjectIdentifierOrKey(SubjectIdentifierOrKeyType.PublicKeyInfo, publicKeyInfo);
+
+                    break;
+
+                default:
+                    throw new CryptographicException(SR.Format(SR.Cryptography_Cms_Invalid_Originator_Identifier_Choice, tag));
+            }
+
+            if (encodedKeyAgreement.PeekTag() == contextSpecific1)
+            {
+                // Skip extra keying material; this one isn't exposed in the API and Framwork does
+                // nothing with it.
+                encodedKeyAgreement.SkipValue();
+            }
+
+            AlgorithmIdentifier keyEncryptionAlgo = encodedKeyAgreement.ReadAlgoIdentifier();
+
+            // As there might be more than one element in RecipientEncryptedKeys, we have to create one KeyAgreeRecipientInfo for each
+            // one of the elements. 
+            List<SubjectIdentifier> recipientIdSet = new List<SubjectIdentifier>();
+            List<CryptographicAttributeObject> otherKeyMaterialSet = new List<CryptographicAttributeObject>();
+            List<DateTime> dateSet = new List<DateTime>();
+            List<byte[]> encryptionKeySet = new List<byte[]>();
```

Comment:
encrypted, not encryption.  An encrypted key must be decrypted to be a key.  An encryption key is ready to go.